azure-auth-helper/src/main/java/com/microsoft/azure/auth/AzureAuthHelper.java

```diff
@@ -39,7 +78,59 @@ public static AzureCredential oAuthLogin(AzureEnvironment env) throws AzureLogin
      * @throws AzureLoginFailureException when there are some errors during login.
      */
     public static AzureCredential deviceLogin(AzureEnvironment env) throws AzureLoginFailureException {
-        throw new UnsupportedOperationException("Not implemented");
+        Object logger = null;
+        Field levelField = null;
+        Object oldLevelValue = null;
+
+        try {
+            System.out.println(AUTH_WITH_DEVICE_LOGIN);
+
+            try {
+                // disable log4j of AuthenticationContext, otherwise the pending user
+                // authorization log
+                // will be print every second.
+                // see
+                // https://github.com/AzureAD/azure-activedirectory-library-for-java/issues/246
+                logger = LoggerFactory.getLogger(AuthenticationContext.class);
+                levelField = logger.getClass().getSuperclass().getDeclaredField("currentLogLevel");
+                oldLevelValue = setObjectValue(logger, levelField, LocationAwareLogger.ERROR_INT + 1);
+            } catch (Exception e) {
+                System.out.println("Failed to disable the log of " + AuthenticationContext.class.getName() + ", it will continue being noisy.");
+            }
+            return new AzureCredentialCallable(baseURL(env), authenticationContext -> {
+                final DeviceCode deviceCode = authenticationContext.acquireDeviceCode(Constants.CLIENT_ID, env.activeDirectoryResourceId(), null).get();
+                System.out.println(String.format(DEVICE_LOGIN_MESSAGE_TEMPLATE, deviceCode.getVerificationUrl(), deviceCode.getUserCode()));
+                long remaining = deviceCode.getExpiresIn();
+                final long interval = deviceCode.getInterval();
+                AuthenticationResult result = null;
+                while (remaining > 0 && result == null) {
+                    try {
+                        remaining -= interval;
+                        Thread.sleep(interval * 1000);
+                        result = authenticationContext.acquireTokenByDeviceCode(deviceCode, null).get();
+                    } catch (ExecutionException e) {
+                        if (e.getCause() instanceof AuthenticationException &&
```

Comment:
why not just catch AuthenticationException? I don't see the below type cast necessary.

Reply:
e.getCause() not e, the AuthenticationException to ExecutionException is caused by converting calling future.get()